pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/extractor/KerberosExtractor.java

```diff
@@ -0,0 +1,53 @@
+package org.pac4j.kerberos.credentials.extractor;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+
+import org.pac4j.kerberos.credentials.KerberosCredentials;
+import org.pac4j.core.context.HttpConstants;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.credentials.extractor.CredentialsExtractor;
+import org.pac4j.core.exception.CredentialsException;
+import org.pac4j.core.exception.DeferredHttpAction;
+import org.pac4j.core.exception.DeferredHttpActionCallback;
+
+/**
+ * To extract Kerberos headers.
+ *
+ * @author Garry Boyce
+ * @since 1.9.1
+ */
+public class KerberosExtractor implements CredentialsExtractor<KerberosCredentials> {
+    private final String clientName;
+
+    public KerberosExtractor(final String clientName) {
+        this.clientName = clientName;
+    }
+
+    @Override
+    public KerberosCredentials extract(WebContext context) {
+        final String header = context.getRequestHeader(HttpConstants.AUTHORIZATION_HEADER);
+        if (header == null) {
+            throw DeferredHttpAction.deferredHttpAction("Kerberos Header not found", new DeferredHttpActionCallback() {

```

Comment:
I feel this is a larger problem.. This isn't just a kerberos problem. For instance lets say I start a Basic authentication and I check for a valid auth in the header and determine it doesn't exist.. If I was using an indirect client then it would ask for credentials and if I was using a direct client it would just fail. And so I'm saying I want flexibility from the client side to decide what behavior I want. i.e provide the client with a callback so the client can decide to continue on with the proposed behavior or if to do it's own thing.

Wrt the filter level what I'm doing specifically is using Spring security filters and I have both a custom direct and indirect entry point where i decide what to do in each case with the security:http spring config pointing to one or the other. And that's the client (or caller). But that client needs the pac4j client to provide the callback which it currently is not doing. Furthermore that client wants the pac4j client to not do things to the response without it's say so. 

I don't see how any of the mechanisms you propose address the flexibility I'm looking for.

Right now in my Direct entry point I have code like this:

> if (isTryToRespond) {
>             if (client instanceof DirectExtended) {
>                 ((DirectExtended) client).handleForbidden(context);
>                 return;
>             } else if (client instanceof DirectBasicAuthClient) {
>                 RequiresHttpAction.unauthorized("Requires authentication", context, realmName);
>                 return;
>             }
>         }

I'd prefer a callback on the client or the exception provided which would allow me to run the client specified code.